File: cpp/lib/matlab_funcs/EstimateStateFCN.cpp
#include "matlab_funcs.h"

#define FILTER_MODE 1
#define RTK 1

Vector13 EstimateStateFCN(Vector13 x_est, t_constantsASTRA constantsASTRA, Vector15 z, float dT, float GND, Matrix12_12 &P, bool new_imu_packet, bool new_gps_packet) {
  //// M-EKF Implementation
  // Filter mode (1 for full INS when GPS signals available, 0 for pure
  // integration after launch when no GPS available | limit flight time)

  // Remove bias from gyro
  z.segment<3>(3) = z.segment<3>(3) - x_est.segment<3>(10) * (FILTER_MODE == 1 || GND == 1);

  // Extract quaternion
  Vector12 dx = Vector12::Zero();
  Vector4 q = x_est.segment<4>(0);
  Vector4 t_zseg;
  t_zseg << 0, z.segment<3>(3);
  Vector4 qdot = 0.5 * HamiltonianProd(q) * t_zseg;
  x_est.segment<4>(0) = q + qdot * dT;
  q = x_est.segment<4>(0);

  // A-priori quaternion estimate and rotation matrix
  q.normalize();
  Matrix3_3 R_b2i = quatRot(q).transpose();

  // State Transition Matrix
  Matrix12_12 F = StateTransitionMat(z.segment<3>(0), z.segment<3>(3), R_b2i);

  Vector3 g_vec3;
  g_vec3 << 0, 0, constantsASTRA.g;

  // Propagate rest of state using IMU
  x_est.segment<3>(7) = x_est.segment<3>(7) + (R_b2i * z.segment<3>(0) - g_vec3) * dT;
  x_est.segment<3>(4) = x_est.segment<3>(4) + x_est.segment<3>(7) * dT;

  // Discrete STM
  Matrix12_12 Phi = matrixExpPade6(F * dT);

  // Extract Matrices
  Matrix12_12 Q = constantsASTRA.Q;
  Matrix6_6 R = constantsASTRA.R;

  // Process Noise Covariance and a-priori propagation step
  Q = 0.4 * Q;
  P = Phi * P * Phi.transpose() + Q;

  if (new_imu_packet && (FILTER_MODE == 1 || GND == 1)) {

    // Measurement matrix
    Matrix6_12 H = Matrix6_12::Zero();
    H.block<3, 3>(0, 0) = zetaCross(R_b2i.transpose() * g_vec3);
    H.block<3, 3>(3, 0) = zetaCross(R_b2i.transpose() * constantsASTRA.mag);

    // Measurement Noise Covariance
    float w = 1 + 1e5 * (1 - GND);
    R.block<3, 3>(0, 0) = R.block<3, 3>(0, 0) * w;

    // A priori covariance and Kalman gain

    Matrix12_6 L = P * H.transpose() * (H * P * H.transpose() + R).inverse();

    // Predicted measurements
    Vector6 z_hat;
    z_hat << R_b2i.transpose() * g_vec3, R_b2i.transpose() * constantsASTRA.mag;

    // Kalman Gain Weighting based on predicted acceleration
    Matrix12_12 ILH = (Matrix12_12::Identity() - L * H);
    P = ILH * P * ILH.transpose() + L * R * L.transpose();
    Vector6 z_slice;
    z_slice << z.segment<3>(0), z.segment<3>(6);
    Vector6 residual = (z_slice - z_hat);
    dx = dx + L * residual;
  }
  if (new_gps_packet && (FILTER_MODE == 1 || GND == 1)) {

    // Measurement matrix
    Matrix6_12 H = Matrix6_12::Zero();
    H.block<3, 3>(0, 3) = Matrix3_3::Identity();
    H.block<3, 3>(3, 6) = Matrix3_3::Identity();

    // Measurement Covariance Matrix
    float gps_pos_covar = 1 * RTK + 10 * (1 - RTK);
    float gps_vel_covar = gps_pos_covar;
    Vector6 R_vec;
    R_vec << gps_pos_covar * gps_pos_covar * Vector3::Ones(), gps_vel_covar * gps_vel_covar * Vector3::Ones();
    Matrix6_6 R = R_vec.asDiagonal();

    // A priori covariance and Kalman gain
    Matrix12_6 L = P * H.transpose() * (H * P * H.transpose() + R).inverse();

    // Predicted measurements
    Vector6 z_hat;
    z_hat << x_est.segment<3>(4), x_est.segment<3>(7);

    // Kalman Gain Weighting based on predicted acceleration
    Matrix12_12 ILH = (Matrix12_12::Identity() - L * H);
    P = ILH * P * ILH.transpose() + L * R * L.transpose();
    Vector6 residual = (z.segment<6>(9) - z_hat);
    Vector12 inn = L * residual;
    dx = dx + inn;
  }

  if (FILTER_MODE == 1 || GND == 1) {
    // Update full-state estimates
    Vector4 dq;
    dq << 1, dx.segment<3>(0) / 2;
    dq.normalize();

    Vector4 q_nom = HamiltonianProd(q) * dq;
    q_nom.normalize();
    x_est.segment<4>(0) = q_nom;
    x_est.segment<9>(4) = x_est.segment<9>(4) + dx.segment<9>(3);
  }
  x_est.segment<6>(4) = x_est.segment<6>(4) * (1 - GND);
  return x_est;
}

File: Filtering/Kalman FIlter/EstimateStateFCN.m
function [x_est, dx] = EstimateStateFCN(x_est,constantsASTRA,z,dT,GND)

%% M-EKF Implementation
% Filter mode (1 for full INS when GPS signals available, 0 for pure
% integration after launch when no GPS available | limit flight time)
FILTER_MODE = 1;

% Remove bias fromIMU
z(1:3) = z(1:3) - x_est(14:16) * (FILTER_MODE == 1 || GND == 1);
z(4:6) = z(4:6) - x_est(11:13) * (FILTER_MODE == 1 || GND == 1);
z(7:9) = z(7:9) - x_est(17:19) * (FILTER_MODE == 1 || GND == 1);

% Extract quaternion
dx = zeros(18,1);
q = x_est(1:4);
qdot = 0.5 * HamiltonianProd(q) * [0; z(4:6)]; 
x_est(1:4) = q + qdot * dT;
q = x_est(1:4);

% A-priori quaternion estimate and rotation matrix
q = q / norm(q);
R_b2i = quatRot(q)';

% Process Covariance Matrix
persistent P lastZ 
if isempty(P)
    P = 1 * eye(18);  
    lastZ = zeros(15,1);
end

% State Transition Matrix (CHANGE!!)
F = StateTransitionMat(z(1:3), z(4:6), R_b2i);

% Propagate rest of state using IMU
x_est(8:10) = x_est(8:10) + (R_b2i * z(1:3) - [0; 0; constantsASTRA.g]) * dT;
x_est(5:7) = x_est(5:7) + x_est(8:10) * dT;

% Discrete STM
Phi = expm(F * dT);

% Extract Matrices
Q = constantsASTRA.Q;
R = constantsASTRA.R;

% Process Noise Covariance and a-priori propagation step
Q = 0.5 * Q;
P = Phi * P * Phi' + Q;
RTK = 0;

if sum(lastZ(1:9) - z(1:9)) ~=0 && (FILTER_MODE == 1 && GND == 1)

    % Measurement matrix
    H = zeros(6,18);
    H(1:3, 1:3) = zetaCross(R_b2i' * [0; 0; constantsASTRA.g]);
    H(1:3, 13:15) = eye(3);
    H(4:6, 1:3) = zetaCross(R_b2i' * constantsASTRA.mag);
    H(4:6, 16:18) = eye(3);

    % Measurement Noise Covariance
    w = 1 + 1e5 * (1 - GND);
    R(1:3,1:3) = R(1:3,1:3) * w;
    
    % A priori covariance and Kalman gain
    L = P * H' / (H * P * H' + R);
    
    % Predicted measurements 
    z_hat = [R_b2i' * [0; 0; constantsASTRA.g];
             R_b2i' * constantsASTRA.mag];
    
    % Kalman Gain Weighting based on predicted acceleration
    ILH = (eye(18) - L * H);
    P = ILH * P * ILH' + L * R * L';
    residual = (z([1:3 7:9]) - z_hat);
    dx = dx + L * residual;
end
if sum(lastZ(10:15) - z(10:15)) ~=0 && (FILTER_MODE == 1 || GND == 1)

    % Measurement matrix
    H = zeros(6,18);
    H(1:3, 4:6) = eye(3);
    H(4:6, 7:9) = eye(3);

    % Measurement Covariance Matrix
    gps_pos_covar = 1 * RTK + 150 * (1 - RTK);
    gps_vel_covar = gps_pos_covar * 0.1;
    R = diag([gps_pos_covar^2 * ones(3,1); gps_vel_covar^2 * ones(3,1)]);

    % A priori covariance and Kalman gain
    L = P * H' / (H * P * H' + R);

    % Predicted measurements 
    z_hat = [x_est(5:7);
             x_est(8:10)];

    % Kalman Gain Weighting based on predicted acceleration
    ILH = (eye(18) - L * H);
    P = ILH * P * ILH' + L * R * L';
    residual = (z(10:15) - z_hat);
    inn = L * residual;
    dx = dx + inn;
end
if FILTER_MODE == 1 || GND == 1
    % Update full-state estimates
    dq = [1; dx(1:3) / 2];
    dq = dq / norm(dq);
    
    q_nom = quatmultiply(q', dq');
    q_nom = q_nom / norm(q_nom); 
    dx(10:18) = dx(10:18) * GND;
    x_est(1:4) = q_nom';
    x_est(5:19) = x_est(5:19) + dx(4:18);
end
x_est(5:10) = x_est(5:10) * (1 - GND);
lastZ = z;
end